proposals/0000-language-editions.rst

```diff
@@ -0,0 +1,1144 @@
+how a language edition can control more than just extensions (importantly, one
+can control *warnings* as well), and proposes several concrete language editions.
+One of the proposals is for ``Stable2024``, which will continue to be supported,
+without change to the set of user programs accepted, for at least 3 years.
+
+This is extracted from a longer `manifesto`_ about how GHC should treat language
+extensions. For some introductory text describing the mindset that leads to this
+design, readers may want to skim the introductory sections of that manifesto.
+However, the effective payload of this proposal is all in this one document.
+
+Motivation
+----------
+
+The primary motivation behind the use of language editions is that they
+can succinctly inform GHC what kind of user it's faced with, so GHC
+can behave accordingly.
+
+Though the details are spelled out below, it's necessary to introduce
+some of the language editions I'm proposing:
+
+* ``Stable2024``: Code compiled in the ``Stable2024`` edition will be
+  expected to compile (assuming stability of libraries) under that
+  same language edition on all GHCs release for 6 years,
+  until the beginning of 2030. (We may end up falling short of this
+  expectation in corner cases, but it will remain a reasonable expectation
+  to have.)
+
+* ``Experimental2024``: Switching to the ``Experimental`` series of
+  editions gives you access to experimental features of GHC, which might
+  reasonably evolve and break existing programs.
+
+* ``Latest``: This language edition is the latest and greatest that GHC
+  has to offer. Compiling with ``Latest`` might break between releases.
+
+* ``Student2024``: The student edition of the language has extra guardrails.
+  Furthermore, any code written in the ``Student`` edition is expected
+  to compile with all future versions of GHC for 10 years, meant to
+  echo the expected lifetime of a textbook.
+
+With that out of the way, here are some scenarios that the language editions
+model helps us to address:
+
+* **Stability**. Once we include a feature in a ``Stable`` edition, it
+  will not change. However, that does not mean GHC cannot evolve. If we
+  identify a ``Stable`` feature that nevertheless deserves an upgrade, we
+  can do so, provided we continue to support the old behavior. So
+  type inference around ``FlexibleInstances``, say, might work one way in
+  ``Stable2024`` and differently in ``Stable2027``. If we release a new
+  ``Stable`` edition every 3 years, users have three years to upgrade before
+  we no longer guarantee support. A three-year timeline is long enough that
+  we might imagine contributors writing upgrade tools in that timeframe.
+
+* **Error messages**. Because the language edition describes the user (very
+  coarsely!) we can tailor error messages to be appropriate for them. The
+  ``Student`` editions will not advise ``DataKinds``. The ``Stable`` editions
+  will not advise ``LinearTypes``.
```

Comment:
Yes, I think that's a reasonable principle, but it has limits:

- With stability, a programmer's decision to adopt an unstable feature affects not only them, but also others, including downstream dependencies. Of course we could say so in an error message, but we easily get into a situation where the person who is negatively impacted (a downstream programmer who wants stability) is not in the position of mitigating the problem. So I think this plan doesn't quite work for stability.
- With beginners, it will be hard to know how to phrase the message in a way that fits what they have learned (because we cannot easily know what they have learned!). Giving a user information that they cannot interpret is not helpful. (Example: describing variables as rigid skolems is accurate, but not very useful unless you have a friend nearby to explain what a skolem is.)

Some of this is moot, as I plan to get rid of the consistently unpopular `Student` edition.

Reply:
> So I think this plan doesn't quite work for stability.

I'm not sure I quite get the concrete example you're thinking of. Here's one. Suppose that I have `Stable2024` set. I then write some code that could work if I enable an unstable language extension. Here's what I would propose to do:
- Say in the error message: "Perhaps you meant X (unstable)?". HLS might offer this as a code action
- If they add it, they get an error saying: "Extension X is unstable and not compatible with your language edition, did you mean to set Unstable?". HLS might offer this as a code action

Here at each point we are giving the user the information that they need, and allowing them to make the decision, or not. If we design the system so that there are further consequences, we can highlight them: "...did you mean to set Unstable (downstream stable modules will not be able to use this module)?"

> describing variables as rigid skolems is accurate, but not very useful unless you have a friend nearby to explain what a skolem is

Jargon is a pervasive problem for sure. Fortunately, we have the GHC error message index, which provides a place to explain errors in more detail, perhaps including the jargon or linking to a place where the jargon is explained.